src/nd_ODE_Static.jl

```diff
@@ -36,42 +37,69 @@ end
 end
 
 
+"""
+    Unroll needs its own function and the loop to unroll needs to iterate
+    over one of its arguments (modulo `length`).
+"""
+@unroll function component_loop!(dx, p, t, gd, gs,
+                              unique_components, unique_c_indices, parallel)
+    @unroll for j in 1:length(unique_components)
+        # @Threads.threads and @generated clash
+        # To make them uninvolved with each other we need another function
+        _inner_loop!(dx, p, t, gd, gs,
+                           unique_components[j], unique_c_indices[j], parallel)
+    end
+end
```

Comment:
Working on the prototyp I've realised that the `@unroll` of the `Tuple` might not be even necessary. If the tuples get really big this isn't that performant any more. But since you introduced the `_inner_loop!` as a function barrier the type instability of the `unique_components[j]` results in a small performance penalty but won't propagate further: `_inner_loop!` is typestable again. I think this is at least worth benchmarking.

Reply:
Only tuple is slighly slower than unrolled for the benchmark suite. There are no really long tuples in that suite however:
| ID                                                   | time ratio                   | memory ratio  |
|------------------------------------------------------|------------------------------|---------------|
| `["diffusion", "ode_edge", "assemble", "1000"]`      |                1.10 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "ode_edge", "assemble", "10000"]`     |                1.07 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "ode_edge", "call_mt", "100"]`        | 0.93 (5%) :white_check_mark: |    1.00 (1%)  |
| `["diffusion", "ode_edge", "call_mt", "1000"]`       | 0.87 (5%) :white_check_mark: |    1.00 (1%)  |
| `["diffusion", "ode_edge", "call_mt", "10000"]`      |                1.26 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "static_edge", "call", "10"]`         | 0.89 (5%) :white_check_mark: |    1.00 (1%)  |
| `["diffusion", "static_edge", "call", "100"]`        |                1.07 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "static_edge", "call", "10000"]`      |                1.08 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "static_edge", "call_mt", "1000"]`    |                1.19 (5%) :x: |    1.00 (1%)  |
| `["diffusion", "static_edge", "call_mt", "10000"]`   | 0.93 (5%) :white_check_mark: |    1.01 (1%)  |
| `["kuramoto", "heterogeneous", "call", "10"]`        |                1.30 (5%) :x: |  Inf (1%) :x: |
| `["kuramoto", "heterogeneous", "call", "100"]`       |                1.08 (5%) :x: |  Inf (1%) :x: |
| `["kuramoto", "heterogeneous", "call", "1000"]`      |                1.06 (5%) :x: |  Inf (1%) :x: |
| `["kuramoto", "heterogeneous", "call", "10000"]`     |                1.06 (5%) :x: |  Inf (1%) :x: |
| `["kuramoto", "heterogeneous", "call_mt", "10"]`     |                1.08 (5%) :x: | 1.03 (1%) :x: |
| `["kuramoto", "heterogeneous", "call_mt", "100"]`    |                1.20 (5%) :x: | 1.03 (1%) :x: |
| `["kuramoto", "heterogeneous", "call_mt", "1000"]`   |                1.12 (5%) :x: | 1.03 (1%) :x: |
| `["kuramoto", "heterogeneous", "call_mt", "10000"]`  |                   1.04 (5%)  | 1.03 (1%) :x: |
| `["kuramoto", "homogeneous", "call", "10"]`          | 0.88 (5%) :white_check_mark: |    1.00 (1%)  |

Reply:
Do you have a hunch how the performance would compare for large tuples?